### ISSUE TITLE
Disambiguating certificates in a PCKS#11 token

### ISSUE THREAD

**Issue by vtgdias** [MatthiasValvekens/pyHanko#25]
It is more like a help building code for sign a pdf document using my USB token (self signed certificate).

The signature on output file is corrupted. Im using the test code bellow with your pdf template (minimal-with-field_signed.pdf):

```
from pyhanko.sign import signers, pkcs11
from pyhanko.sign import timestamps
from pyhanko.pdf_utils.incremental_writer import IncrementalPdfFileWriter

PKCS11LIB='C:\\Windows\\System32\\eTPKCS11.dll'
SERIALTOKEN=b'xxxxxxxx'
LABEL='xxxxxxxxxxx'
PIN='xxxxxxxxxx'
EXTENSION=".pdf"
DOCUMENT='C:\\Users\\vinicius\\Desktop\\minimal-with-field'
DOCUMENT_EXT=DOCUMENT+EXTENSION
SIGNED_FILE = DOCUMENT+"_signed.pdf"

cms_session = pkcs11.open_pkcs11_session(PKCS11LIB,user_pin=PIN,token_label=LABEL)
cms_signer = pkcs11.PKCS11Signer(cms_session,'')

with open(DOCUMENT_EXT, 'rb') as doc:
    w = IncrementalPdfFileWriter(doc)
    print(w.document_id)
    with open(SIGNED_FILE, 'wb+') as file_out:    
        signers.PdfSigner(signers.PdfSignatureMetadata(field_name='Sig1'),signer=cms_signer).sign_pdf(w,output=file_out)

 ```

The code is able to get my certificate data, as the signature (visible text) shows without any trouble, but the signature comes corrupted as shown bellow:

![sign_error](https://user-images.githubusercontent.com/237875/127948319-dcec852a-3265-49fd-8d7c-0f5c991991af.png)

btw, nice job with this library! tks in advance

**Comment by vtgdias** [comment]
I realized the signature were signed as 'SPAI ICCE Intermediate CA' and should have been my user (vinicius). both cert_label (friendly name) are empty (as shown in Safenet authentication client tools). Is it possible to select certificate by id?

**Comment by MatthiasValvekens** [comment]
Interesting... I briefly considered allowing for querying objects by ID when I wrote that class; I was aware of the fact that many PKCS#11 implementations don't actually enforce uniqueness of labels, but also figured that object IDs might not be stable enough to be useful in practice, so I decided against supporting querying by ID at the time.

I'll see what I can do to work this into the high-level API, but `python-pkcs11` exposes `ID` and `OBJECT_ID` attributes. Presumably at least one of those can be made to do what you want, but I need to look into it.
There might be a cleaner way to handle key/certificate discovery, that doesn't rely on knowing object IDs ahead of time. I'm not super familiar with PKCS#11, so I'd have to look it up.

In the meantime, you can work around the problem if you know what the certificate you need looks like. [This line](https://github.com/MatthiasValvekens/pyHanko/blob/42d9db605bbdb3e22da9b6d537d8aa3f010c385d/pyhanko/sign/pkcs11.py#L274) in `_load_objects()` is responsible for selecting the signer's certificate. So if you override that method in `PKCS11Signer` and insert your own logic to pull the signer's certificate, you're good. You can either pull all certificates and filter them to retrieve the one you need (see also the `cert_registry` property), or replace the `_pull_cert()` call with a more targeted query using `python-pkcs11`.

**Comment by MatthiasValvekens** [comment]
I pushed a commit (see above) that makes it easier to pass in the signer's certificate when the `PKCS11Signer` is instantiated. The `cert_label` parameter is now optional. Instead, you can pass in the certificate directly via the new `signing_cert` parameter, and use the `key_label` parameter to specify the private key handle. This will skip the signer cert lookup on the token. In fact, using this method, the signer's certificate doesn't have to be present in the PKCS#11 device at all.

While this doesn't solve your problem completely, you can now export the certificate from the token using any PKCS#11 client tool and load it out of band. Would that work for you? I'm working under the assumption that looking up object IDs in your token would require probing using a PKCS#11 management utility anyhow, so perhaps this strategy requires less work on your part.

I'll leave this issue open for the time being, since the linked commit doesn't fix everything.

**Comment by vtgdias** [comment]
@MatthiasValvekens,

I was able to export my certificate and read it from python (as1ncrypto.x509), all went fine with `signing_cert` but `key_label` became a issue where i couldnt get it working with an empty key label (yeah key label is also empty), returning `pkcs11.exceptions.PKCS11Error: Could not determine private key handle.` error. So i tried to change the attribute in function call `get_objects()` from pkcs11.py, and even call a differente function `get_key()`, but they all returns me a `TypeError: string argument without an encoding` error. Tried to change to bytes, but no lucky as well.

with code bellow i can check the id of my key (and it is the one i use in `get_key` or `get_objects`):

```
for cert in cms_session.get_objects({Attribute.CLASS: ObjectClass.PRIVATE_KEY}):
            print(cert[Attribute.ID])
```
I am probably missing something stupid...

I was not able to debug it further into pkcs11 lib. 

BTW, if i can get this to work (get_key or get_objects), it is perfect as far as i am concerned! tks for your help and rapid response.

**Comment by MatthiasValvekens** [comment]
Hmm, I'm assuming that there's more than one private key in your token?

What are you seeing when you query `get_key(id=<ID as 'bytes'>, object_class=ObjectClass.PRIVATE_KEY)`? That would seem like it should always return a unique result. I'm not that well-acquainted with the internals of `python-pkcs11`, but a stack trace might help.
Note: in `python-pkcs11`, key IDs are given as `bytes` objects, while labels are `str` objects. Perhaps that's the problem, if you're passing in both.

**Comment by vtgdias** [comment]
I was able to get it working using `get_objects()` with `Attribute.ID` (had to pass `label` as '' too)! Tks for your help.

**Comment by MatthiasValvekens** [comment]
I added a by-ID selection mechanism in the commit linked above, which should fix this issue in the public API as well. Thanks for reporting back on what made it work for you!

PS: I learned that even IDs don't necessarily uniquely identify an object in PKCS#11. For example, on the YubiKey that I experimented with, every key generated on-device has two certificates associated with it with the same ID (a "regular" cert, and an attestation cert). So in some situations you might indeed have to pass both the label and the ID.
